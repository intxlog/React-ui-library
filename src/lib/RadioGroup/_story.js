import React from 'react'
import { storiesOf } from '@storybook/react'
import RadioGroup from './'
import RadioButton from './../RadioButton'

storiesOf('RadioGroup', module)
.add('default', () => (
  <RadioGroup
    name={`example`}
  >
    <RadioButton
      labelText={`Option 1`}
      value={1}
    ></RadioButton>
    <RadioButton
      labelText={`Option 2`}
      value={2}
    ></RadioButton>
  </RadioGroup>
))
.add('required', () => (
  <RadioGroup
    name={`example5`}
    required={true}
    formSubmitted={false}
    isValid={(bool) => {console.log(bool)}}
  >
    <RadioButton
      labelText={`Option 1`}
      value={1}
    ></RadioButton>
    <RadioButton
      labelText={`Option 2`}
      value={2}
    ></RadioButton>
  </RadioGroup>
))
.add('disabled', () => (
  <RadioGroup
    name={`example2`}
    disabled
  >
    <RadioButton
      labelText={`Option 1`}
      value={1}
    ></RadioButton>
    <RadioButton
      labelText={`Option 2`}
      value={2}
    ></RadioButton>
  </RadioGroup>
))
.add('horizontal layout', () => (
  <RadioGroup
    name={`example3`}
    direction={`horizontal`}
  >
    <RadioButton
      labelText={`Option 1`}
      value={1}
    ></RadioButton>
    <RadioButton
      labelText={`Option 2`}
      value={2}
    ></RadioButton>
  </RadioGroup>
))
.add('error', () => (
  <RadioGroup
    name={`example4`}
    error={true}
  >
    <RadioButton
      labelText={`Option 1`}
      value={1}
    ></RadioButton>
    <RadioButton
      labelText={`Option 2`}
      value={2}
    ></RadioButton>
  </RadioGroup>
))
.add('default checked', () => (
  <RadioGroup
    name={`example4`}
  >
    <RadioButton
      labelText={`Option 1`}
      value={1}
      ></RadioButton>
    <RadioButton
      labelText={`Option 2`}
      value={2}
      defaultChecked
    ></RadioButton>
  </RadioGroup>
))