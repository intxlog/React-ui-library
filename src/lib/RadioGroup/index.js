import React from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'

//import css
import styles from './styles.module.scss'

// logic behind all the different types of inputs
class RadioGroup extends React.Component {
  state = {
    value: this.props.defaultValue !== undefined && this.props.defaultValue !== null ? this.props.defaultValue : null,
    error: false,
    isValid: false,
    refreshKey: 1
  }

  componentDidMount() {
    this.reportValidity()

    if (this.props.disabled) {
      this.disableComponent()
    }

    if (this.state.value !== null) {
      this.setState({
        isValid: true
      })
    }
  }

  componentDidUpdate(prevProps, prevState) {
    //when isValid changes report the value
    if (prevState.isValid !== this.state.isValid) {
      this.reportValidity()

      if (this.props.formSubmitted) {
        this.setState({
          error: !this.state.isValid
        })
      }
    }

    if (prevState.value === null && this.state.value !== null) {
      this.setState({
        isValid: true
      })
    }

    //check to see if the formSubmitted prop becomes true
    if (this.props.required) {
      if (this.props.formSubmitted !== prevProps.formSubmitted && this.props.formSubmitted) {
        if (!this.state.isValid) {
          this.setState({
            error: true
          })
        }
      }
    }

    //check to see if the component became disabled
    if (!prevProps.disabled && this.props.disabled) {
      this.disableComponent()
    }

    if (prevProps.disabled && !this.props.disabled) {
      this.enableComponent()
    }
  }

  handleOnChange = (e) => {
    const val = e.target.value
    this.setState({
      value: isNaN(val) ? val : parseInt(val)
    })
    this.props.onChange(val)
  }

  //method to report whether of not the input is valid or not
  reportValidity = () => {
    this.props.isValid(this.state.isValid)
  }

  //method to handle when the component becomes disabled
  disableComponent = () => {
    let keyValue = this.state.refreshKey
    this.setState({
      isValid: true,
      refreshKey: keyValue + 1
    })
  }

  enableComponent = () => {
    let keyValue = this.state.refreshKey
    let isValid = true

    if (this.state.value === null) {
      isValid = false
    }

    this.setState({
      refreshKey: keyValue + 1,
      isValid
    })
  }

  
  render(){
    const children = React.Children.map(this.props.children, (child, index) => {
      let props = {}
      
      //make sure we only pass props and methods to children with a display name of RadioButton
      if (child.type.displayName === `RadioButton`) {
        props = {
          key: `${this.state.refreshKey}`,
          id: `${this.props.name}${index}`,
          error: this.state.error || this.props.error,
          name: this.props.name,
          disabled: this.props.disabled,
          onChange: this.handleOnChange,
          defaultValue: this.state.value
        }
      }
      
      return React.cloneElement(child, props)
    })
    
    
    return (
      <div className={styles.container}>
        <div className={classNames({
          [styles.childrenWrapper]: true,
          [this.props.className] :  this.props.className,
          [styles.horizontal]: this.props.direction === `horizontal`
        })}>
          {children}
        </div>
      </div>
    )
  }
}

//rules for props being passed in
RadioGroup.propTypes = {
  required: PropTypes.bool,
  name: PropTypes.string.isRequired,
  disabled: PropTypes.bool,
  onChange: PropTypes.func,
  isValid: PropTypes.func,
  className: PropTypes.object,
  direction: PropTypes.oneOf([`horizontal`, `vertical`]).isRequired,
  defaultValue: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
    PropTypes.bool
  ])
}

// Specifies the default values for props:
RadioGroup.defaultProps = {
  direction: `vertical`,
  onChange: () => {},
  isValid: () => {}
}

//maintain the name for documentation purposes
RadioGroup.displayName = 'RadioGroup'

export default RadioGroup