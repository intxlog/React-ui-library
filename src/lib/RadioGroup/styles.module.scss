.container {
  margin-bottom: 16px;
}

.childrenWrapper {
  display: grid;
  grid-row-gap: 10px;

  &.horizontal {
    justify-content: start;
    grid-template-columns: repeat(auto-fill, minmax(100px, 1fr))
  }
}