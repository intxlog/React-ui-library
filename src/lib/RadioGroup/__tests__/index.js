import React from 'react'
import { shallow } from 'enzyme'
import RadioGroup from '..'

describe('RadioGroup', () => {
  //mock functions
  const onChangeMock = jest.fn()
  const isValidMock = jest.fn()

  let props = {
    name: `example`,
    onChange: onChangeMock,
    isValid: isValidMock
  }
  
  //create a shallow wrapper that can be reused
  let wrapper
  //initialize the components
  beforeEach(() => {
    wrapper = shallow(<RadioGroup {...props}/>)
  })

  //reset the mocks after each test
  afterEach(() => {
    onChangeMock.mockReset()
    isValidMock.mockReset()
  })

  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot()
  })

  describe('when formSubmitted prop changes and becomes true', () => {
    it('sets the correct state', () => {
      wrapper.setProps({
        formSubmitted: true
      })

      expect(wrapper.state().error).toEqual(true)
    })
  })

  //Methods
  describe('when handleOnChange is called', () => {
    it('calls the correct function', () => {
      const e = {
        target: {
          value: 'test'
        }
      }

      wrapper.instance().handleOnChange(e)
      expect(onChangeMock).toHaveBeenCalled()
      expect(onChangeMock).toHaveBeenCalledWith('test')
    })
    it('sets the correct state', () => {
      const e = {
        target: {
          value: 'test'
        }
      }

      wrapper.instance().handleOnChange(e)
      expect(wrapper.state().value).toEqual('test')
    })
  })
})