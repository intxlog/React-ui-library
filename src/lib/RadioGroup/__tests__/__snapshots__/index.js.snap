// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioGroup renders correctly 1`] = `
<div
  className="container"
>
  <div
    className="childrenWrapper"
  />
</div>
`;
